pkg/lib/operatorstatus/monitor.go

```diff
@@ -0,0 +1,221 @@
+
+	// If there is an upgrade going on, Final is set to the latest version of
+	// the CSV that is replacing current.
+	// For a chain like this, Current (v1) -> v2 -> v3 -> Final (v4)
+	// Final will be set to the CSV linked to v4.
+	// Otherwise this value is set to nil.
+	Final *v1alpha1.ClusterServiceVersion
+
+	// Deleted indicates whether Current has been deleted.
+	Deleted bool
+}
+
+// Monitor is an interface that wraps the Run method.
+//
+// Run runs forever, it reads from an underlying notification channel and
+// updates an clusteroperator resource accordingly.
+// If the specified stop channel is closed the loop must terminate gracefully.
+type Monitor interface {
+	Run(stopCh <-chan struct{})
+}
+
+// Sender is an interface that wraps the Send method.
+//
+// Send can be used to send notification(s) to the underlying monitor. Send is a
+// non-blocking operation.
+// If the underlying monitor is not ready to receive the notification will be lost.
+// If the notification context specified is nil then it is ignored.
+type Sender interface {
+	Send(context *NotificationContext)
+}
+
+func (n *NotificationContext) String() string {
+	replaces := "<nil>"
+	if n.Final != nil {
+		replaces = n.Final.GetName()
+	}
+
+	return fmt.Sprintf("name=%s csv=%s deleted=%s replaces=%s", n.Name, n.Current.GetName(), strconv.FormatBool(n.Deleted), replaces)
+}
+
+type monitor struct {
+	notificationCh chan *NotificationContext
+	writer         *Writer
+	logger         *logrus.Entry
+	reporter       *reporter
+	names          []string
+}
+
+func (m *monitor) Send(context *NotificationContext) {
+	if context == nil {
+		return
+	}
+
+	select {
+	case m.notificationCh <- context:
+	default:
+		m.logger.Warn("monitor not ready to receive")
+	}
+}
+
+func (m *monitor) Run(stopCh <-chan struct{}) {
+	m.logger.Info("starting clusteroperator monitor loop")
+	defer func() {
+		m.logger.Info("exiting from clusteroperator monitor loop")
+	}()
+
+	// First, we need to ensure that cluster operator API is available.
+	// We will keep probing until it is available.
+	for {
+		exists, err := m.writer.IsAPIAvailable()
+		if err != nil {
+			m.logger.Infof("ClusterOperator api not present, skipping update (%v)", err)
+		}
+
+		if exists {
+			m.logger.Info("ClusterOperator api is present")
+			break
+		}
+
+		// Wait before next probe, or quit if parent has asked to do so.
+		select {
+		case <-time.After(defaultProbeInterval):
+		case <-stopCh:
+			return
+		}
+	}
+
+	// If we are here, cluster operator is available.
+	// We are expecting CSV notification which may or may not arrive.
+	// Given this, let's write an initial ClusterOperator object with our expectation.
+	m.logger.Infof("initializing clusteroperator resource(s) for %s", m.names)
+
+	for _, name := range m.names {
+		if err := m.init(name); err != nil {
+			m.logger.Errorf("initialization error - %v", err)
+			break
+		}
+	}
+
+	for {
+		select {
+		case context := <-m.notificationCh:
```

Comment:
It almost seems like `monitor` itself is another operator. I wonder what this would look like using the `QueueInformer` abstractions. Potentially, this could be another `SyncHandler` registered to the CSV `QueueInformer`. When a `SyncHandler` fails for an `Operator`, the sync is "re-queued" and tried again (up to a certain point). Could that be useful here?